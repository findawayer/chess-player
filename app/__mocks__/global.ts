import type { ReactNode } from 'react';

// Mock `window.matchMedia`
// https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Mock next/link with prefetch.
// https://github.com/vercel/next.js/issues/16864#issuecomment-702069418
jest.mock('next/link', () => ({ children }: { children: ReactNode }) =>
  children,
);

jest.mock('next/router', () => ({
  useRouter() {
    return {
      basePath: '',
      pathname: '/',
      route: '/',
      asPath: '/',
      query: {},
      push: jest.fn(),
      replace: jest.fn(),
      reload: jest.fn(),
      back: jest.fn(),
      prefetch: jest.fn().mockResolvedValue(undefined),
      beforePopState: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn(),
        emit: jest.fn(),
      },
      isFallback: false,
    };
  },
}));

export {};
