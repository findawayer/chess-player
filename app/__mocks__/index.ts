export * from './next/router';
export * from './user';
