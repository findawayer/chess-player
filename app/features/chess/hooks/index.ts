export * from './useChessClock';
export * from './useChessGuides';
export * from './useChessPromotion';
export * from './useChessSettings';
export * from './useChessValidator';
export * from './useStockfish';
