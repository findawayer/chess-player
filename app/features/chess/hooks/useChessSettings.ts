import { ChessBoardColor } from '@prisma/client';
import { useEffect, useRef, useState } from 'react';

import {
  DEFAULT_AUTO_QUEEN,
  DEFAULT_BOARD_COLOR,
  DEFAULT_SHOW_LEGAL,
  DEFAULT_SHOW_RECENT,
} from '~app/features/chess/constants';
import { CurrentUser } from '~app/graphql';

export interface ChessSettings {
  autoQueen: boolean;
  boardColor: ChessBoardColor;
  showLegal: boolean;
  showRecent: boolean;
}

/** Map user settings about chess gaming from the current auth user payload. */
export const createChessSettings = (me: CurrentUser | null): ChessSettings => ({
  autoQueen: me?.chessAutoQueen || DEFAULT_AUTO_QUEEN,
  boardColor: me?.chessBoardColor || DEFAULT_BOARD_COLOR,
  showLegal: me?.chessShowLegal || DEFAULT_SHOW_LEGAL,
  showRecent: me?.chessShowRecent || DEFAULT_SHOW_RECENT,
});

/** Convert chess settings to comply with database table keys. (preefix `chess-`) */
export const mapChessSettings = (settings: ChessSettings) => ({
  chessAutoQueen: settings.autoQueen,
  chessBoardColor: settings.boardColor,
  chessShowLegal: settings.showLegal,
  chessShowRecent: settings.showRecent,
});

export const useChessSettings = (me: CurrentUser | null) => {
  // Chess settings from currently authenticated user.
  const [settings, setSettings] = useState(createChessSettings(me));
  // This prevents refetching settings from server after submitting new data.
  const isHydrated = useRef(false);

  /** Change a single entry in settings object. */
  const changeSetting = (
    name: keyof ChessSettings,
    value: ChessSettings[keyof ChessSettings],
  ): void => {
    setSettings(previousSettings => ({
      ...previousSettings,
      [name]: value,
    }));
  };

  // Refresh chess settings when the user authentication status changes.
  useEffect(() => {
    // Prevent duplicate call.
    if (!isHydrated.current && me) {
      isHydrated.current = true;
      setSettings(createChessSettings(me));
    }
  }, [me]);

  return [settings, changeSetting] as const;
};
