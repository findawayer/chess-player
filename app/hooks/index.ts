export * from './useColorMode';
export * from './useUser';
export * from './useVisibilityToggle';
