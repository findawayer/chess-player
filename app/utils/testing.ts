import type { User } from '@prisma/client';

const fakeTime = new Date();

export const fakeUser = (overrides?: Partial<User>): User => ({
  id: 1,
  email: 'some@email.com',
  password: '1234',
  name: 'someone',
  role: 'ADMIN',
  createdAt: fakeTime,
  updatedAt: fakeTime,
  verified: false,
  resetToken: null,
  resetTokenExpiry: null,
  colorMode: null,
  chessAutoQueen: null,
  chessBoardColor: null,
  chessShowLegal: null,
  chessShowRecent: null,
  ...overrides,
});
