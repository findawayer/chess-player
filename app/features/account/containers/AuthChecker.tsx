import { Alert } from '@material-ui/lab';
import type { UserRole } from '@prisma/client';
import React from 'react';

import PleaseSignin from '~app/features/account/components/PleaseSignin';
import PleaseSigninInline from '~app/features/account/components/PleaseSigninInline';
import { useUser } from '~app/hooks/useUser';
import { hasRole } from '~app/utils/roles';

interface AuthCheckerProps {
  children: React.ReactNode;
  inline?: boolean;
  requiredRole?: UserRole;
}

/**
 * Prevent anon users from a certain page or content. Wrap the content
 * with this component.
 *
 * @example
 *   <AuthChecker>
 *     <Admin />
 *   </AuthChecker>

 *   <AuthChecker>
 *     Any React node can be passed as child node.
 *   </AuthChecker>
 */
const AuthChecker: React.FC<AuthCheckerProps> = ({
  children,
  inline,
  requiredRole,
}) => {
  const me = useUser();
  // Not logged in!
  if (!me) {
    return inline ? <PleaseSigninInline /> : <PleaseSignin />;
  }
  // User is missing required role.
  if (requiredRole && !hasRole(me, requiredRole)) {
    return (
      <Alert severity="error" variant="filled">
        You do not have access to this page.
      </Alert>
    );
  }
  // Good to go :)
  return <>{children}</>;
};

export default AuthChecker;
