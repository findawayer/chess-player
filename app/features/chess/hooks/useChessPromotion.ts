import update from 'immutability-helper';
import { useCallback, useEffect, useState } from 'react';

import { stringifySquare } from '~app/features/chess/utils';
import type {
  ChessPromotion,
  ChessPromotionVariant,
} from '~app/features/chess/types';

/**
 * Create a confirmation prompt to let the player decide which piece to
 * promote their promoting pawn.
 *
 * @returns Array of:
 * - promotion: Chess promotion object. `null` means there is no active promotion.
 * - prompt: Create the confirmation prompt.
 * - select: Confirm the promotion variant.
 * - abort: Dismiss the confirmation prompt.
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useChessPromotion = () => {
  const [promotion, setPromotion] = useState<ChessPromotion | null>(null);

  /**
   * Open a confirm dialog to let the player decide
   * which piece to promote the pawn to.
   */
  const promptPromotion = useCallback(
    ({ color, from, to }: Omit<ChessPromotion, 'variant'>) => {
      setPromotion({
        color,
        from: stringifySquare(from),
        to: stringifySquare(to),
      });
    },
    [],
  );
  /** Select the piece variant to promote the current pawn to. */
  const selectPromotion = useCallback(
    (variant: ChessPromotionVariant) => {
      if (promotion) {
        setPromotion(update(promotion, { $merge: { variant } }));
      }
    },
    [promotion],
  );
  /** Close the promotion dialog without doing anything. */
  const abortPromotion = useCallback(() => {
    setPromotion(null);
  }, []);

  // Reset the promotion status on unmount.
  useEffect(() => {
    return () => setPromotion(null);
  }, []);

  return { promotion, promptPromotion, selectPromotion, abortPromotion };
};
