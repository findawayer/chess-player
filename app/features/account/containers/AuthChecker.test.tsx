import { MockedProvider } from '@apollo/client/testing';
import { render, waitFor } from '@testing-library/react';

import { mockUser } from '~app/__mocks__';
import { UserProvider } from '~app/hooks';
import { CURRENT_USER_QUERY } from '~app/graphql';
import AuthChecker from './AuthChecker';

describe('<AuthChecker />', () => {
  it('renders a please sign in dialog to anonymous users.', async () => {
    const anonymousMocks = [
      {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: null } },
      },
    ];
    const { container } = render(
      <MockedProvider mocks={anonymousMocks} addTypename={false}>
        <UserProvider>
          <AuthChecker>ðŸ’©</AuthChecker>
        </UserProvider>
      </MockedProvider>,
    );

    await waitFor(() => 0);
    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent('Please sign in');
  });

  it('renders the child React node when the user is signed in.', async () => {
    const authUserMocks = [
      {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: mockUser() } },
      },
    ];
    const { container } = render(
      <MockedProvider mocks={authUserMocks} addTypename={false}>
        <UserProvider>
          <AuthChecker>
            <p>ðŸ’©</p>
          </AuthChecker>
        </UserProvider>
      </MockedProvider>,
    );

    await waitFor(() => 0);
    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent('ðŸ’©');
  });
});
