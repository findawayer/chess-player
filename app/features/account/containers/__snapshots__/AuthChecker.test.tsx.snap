// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthChecker /> renders a please sign in dialog to anonymous users. 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root"
    >
      <h2
        class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom MuiTypography-alignCenter"
      >
        Please sign in before continuing.
      </h2>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Get started
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Sign in
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`<AuthChecker /> renders the child React node when the user is signed in. 1`] = `
<div>
  <p>
    ðŸ’©
  </p>
</div>
`;
